docs: allow multiple KMS keys to create CMEK database/backup

hwin16 opened the pull request:
Add `kms_key_names` field to create database/create backup code snippets.

BEGIN_COMMIT_OVERRIDE
docs: allow multiple KMS keys to create CMEK database/backup
END_COMMIT_OVERRIDE

hwin16 commented:
I am a googler following self-serve guide.

harshachinta commented on google/cloud/spanner_v1/backup.py:
```diff
@@ -280,16 +280,24 @@ def create(self):
             (
                 self._encryption_config
-                and self._encryption_config.kms_key_name
+                and (
```
@hwin16 Admin features will directly use the autogenerated code and will not need any handwritten logic. So we can remove these changes in backup.py, database.py and test_database.py.

hwin16 committed:
Removed api files

hwin16 commented:
Thank you @harshachinta! PTAL.

harshachinta committed:
Merge branch 'main' into main

harshachinta commented:
@hwin16 Can you please fix the lint by running `nox -s blacken` in samples/samples folder

hwin16 committed:
Fixed lint errors

hwin16 committed:
Merge branch 'main' of https://github.com/hwin16/python-spanner

hwin16 commented:
@harshachinta , I have fixed the linting issue. I fixed some issues with integration tests but wasn't able to run them. Our feature is adding repeated field to proto and similar to existing unit test "create_database_with_encryption_key". When I run  existing unit test, that is failing without any changes. 

![Screenshot 2024-09-05 at 2 22 20 PM](https://github.com/user-attachments/assets/f9da2ec6-f865-4712-9f53-60b1a33e6e2b)

harshachinta commented:
@hwin16 
This test is dependent on another test that creates the instance once. The dependency annotation is missing for these tests. Let me fix it in seperate PR.
However you can run the following command to get it working,
```
nox -s py-3.11 -- -k "test_create_database_explicit or test_create_database_with_encryption_config"
```

hwin16 committed:
Fixed integration tests.

hwin16 committed:
Fixed lint in snippets_test.py

hwin16 committed:
Merge branch 'main' into main

hwin16 commented:
@harshachinta , done fixing integration tests and they are passing now. Currently, skipped them before feature becomes public. 

Passing test results: 
<img width="1439" alt="Screenshot 2024-09-11 at 1 52 52 PM" src="https://github.com/user-attachments/assets/6896b6cb-1314-4561-a75f-6d2aa60a0564">

harshachinta commented on samples/samples/snippets_test.py:
```diff
@@ -160,14 +160,6 @@ def test_create_instance_explicit(spanner_client, create_instance_id):
     retry_429(instance.delete)()
 
 
-def test_create_database_explicit(sample_instance, create_database_id):
```
Why is this code deleted?

hwin16 replied:
Unintentional. Re-added.

harshachinta commented on samples/samples/backup_sample.py:
```diff
@@ -613,6 +697,55 @@ def copy_backup(instance_id, backup_id, source_backup_path):
+    # Create a backup object and wait for copy backup operation to complete.
+    expire_time = datetime.utcnow() + timedelta(days=14)
+    request = backup_pb.CopyBackupRequest(
+        parent=database_admin_api.instance_path(spanner_client.project, "hkw-nam3"),
```
this might have been hardcoded for some testing, can we change it to the variable?

hwin16 replied:
Done.

harshachinta commented on samples/samples/conftest.py:
```diff
@@ -240,8 +240,7 @@ def database_ddl():
 
-@pytest.fixture(scope="module")
-def sample_database(
+def create_sample_database(
```
Why did we make this change in sample_database?

harshachinta replied:
AFAIU, we are creating a helper method to be reused across other methods for creating a database.
PLMK if there is some other reasoning for this change.

hwin16 replied:
exactly. existing `sample_database` is a fixture with regional-instance as its argument. We need multi-region instance(nam3) to test our feature. Thus, refactored.

harshachinta commented on samples/samples/backup_sample.py:
```diff
@@ -119,6 +116,53 @@ def create_backup_with_encryption_key(
 
 # [END spanner_create_backup_with_encryption_key]
 
+# [START spanner_create_backup_with_MR_CMEK]
```
FYI: These sample tags should be same across all the languages.

hwin16 replied:
Yes they are, https://screenshot.googleplex.com/435YU45G4RJYmzW

hwin16 committed:
Resolved comments from reviewer

hwin16 committed:
Merge branch 'main' of https://github.com/hwin16/python-spanner

harshachinta commented on samples/samples/backup_sample_test.py:
```diff
@@ -93,6 +93,45 @@ def test_create_backup_with_encryption_key(
+@pytest.mark.dependency(name="create_backup_with_multiple_kms_keys")
+def test_create_backup_with_multiple_kms_keys(
+    capsys,
+    multi_region_instance,
```
Do we have a fixture for this multi_region_instance?
Where is this instance getting create? I was not able to find in this code

surbhigarg92 committed:
Merge branch 'main' into main

harshachinta committed:
chore: skip tests since KMS keys are not added to test project

harshachinta committed:
Merge branch 'main' into main
